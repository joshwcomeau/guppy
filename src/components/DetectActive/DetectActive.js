// @flow
import React, { Component } from 'react';

type Props = {
  className: string,
  children: (isActive: boolean, isHovered: boolean) => React$Node,
};

type State = {
  isActive: boolean,
  isHovered: boolean,
};

class DetectActive extends Component<Props, State> {
  static defaultProps = {
    className: '',
  };

  state = {
    isActive: false,
    isHovered: false,
  };

  handleMouseDown = (ev: SyntheticEvent<*>) => {
    this.setState({ isActive: true });
  };

  handleMouseUp = (ev: SyntheticEvent<*>) => {
    this.setState({ isActive: false });
  };

  handleMouseOver = (ev: SyntheticEvent<*>) => {
    this.setState({ isHovered: true });
  };

  handleMouseLeave = (ev: SyntheticEvent<*>) => {
    this.setState({ isActive: false, isHovered: false });
  };

  render() {
    const { className } = this.props;
    return (
      <span
        className={className}
        onMouseDown={this.handleMouseDown}
        onMouseUp={this.handleMouseUp}
        onMouseOver={this.handleMouseOver}
        onMouseLeave={this.handleMouseLeave}
      >
        {this.props.children(this.state.isActive, this.state.isHovered)}
      </span>
    );
  }
}

export default DetectActive;
