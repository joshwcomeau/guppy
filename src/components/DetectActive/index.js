// @flow
export { default } from './DetectActive';
