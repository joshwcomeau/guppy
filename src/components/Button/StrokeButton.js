// @flow
import React, { Component } from 'react';
import styled from 'styled-components';

import { COLORS, GRADIENTS } from '../../constants';

import DetectActive from '../DetectActive';
import ButtonBase from './ButtonBase';

export type Props = {
  fillColor?: string,
  strokeColors?: Array<string>,
  showStroke?: boolean,
  children: React$Node,
};

class StrokeButton extends Component<Props> {
  static defaultProps = {
    fillColor: COLORS.lightBackground,
    strokeColors: GRADIENTS.primary,
    showStroke: true,
  };

  render() {
    const {
      fillColor,
      strokeColors,
      showStroke,
      children,
      ...delegated
    } = this.props;

    return (
      <DetectActive>
        {isActive => (
          <Wrapper>
            <Foreground>
              <ButtonBase background={fillColor} {...delegated}>
                <span style={{ display: 'block' }}>{children}</span>
              </ButtonBase>
            </Foreground>

            <Background
              colors={strokeColors}
              isVisible={showStroke}
              isActive={isActive}
            />
          </Wrapper>
        )}
      </DetectActive>
    );
  }
}

const Wrapper = styled.div`
  position: relative;
  display: inline-block;
  padding: 2px;
`;

export const Foreground = styled.span`
  display: block;
  position: relative;
  z-index: 1;
`;

export const Background = styled.div`
  position: absolute;
  z-index: 0;
  top: ${props => (props.isActive ? -2 : 0)}px;
  left: ${props => (props.isActive ? -2 : 0)}px;
  right: ${props => (props.isActive ? -2 : 0)}px;
  bottom: ${props => (props.isActive ? -2 : 0)}px;
  border-radius: 100px;
  background: linear-gradient(25deg, ${props => props.colors.join(', ')});
  opacity: ${props => (props.isVisible ? 1 : 0)};
  transition: opacity 350ms;
`;

export default StrokeButton;
