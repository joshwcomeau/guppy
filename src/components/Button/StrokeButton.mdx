---
name: StrokeButton
menu: Reusable Components
---

import { Playground, Props } from 'docz';
import { Toggle as ToggleState } from 'react-powerplug';

import { RAW_COLORS } from '../../constants';
import StrokeButton from './StrokeButton';

# StrokeButton

<Props of={StrokeButton} />

## Basic usage

<Playground>
  <StrokeButton>Default</StrokeButton>
</Playground>

## Sizes

<Playground>
  <StrokeButton size="xsmall">
    Extra-small button
  </StrokeButton>
</Playground>

<Playground>
  <StrokeButton size="small">
    Small button
  </StrokeButton>
</Playground>

<Playground>
  <StrokeButton size="medium">
    Medium button
  </StrokeButton>
</Playground>

<Playground>
  <StrokeButton size="large">
    Large button
  </StrokeButton>
</Playground>

## Colors

<Playground>
  <StrokeButton strokeColors={[RAW_COLORS.green[500], RAW_COLORS.green[500]]}>
    Solid green
  </StrokeButton>
</Playground>

<Playground>
  <StrokeButton strokeColors={[RAW_COLORS.orange[500], RAW_COLORS.pink[300]]}>
    Orange to pink
  </StrokeButton>
</Playground>

<Playground>
  <StrokeButton strokeColors={[RAW_COLORS.teal[700], RAW_COLORS.lime[500]]}>
    Teal to lime
  </StrokeButton>
</Playground>

## Disabled

<Playground>
  <StrokeButton disabled>
    StrokeButton
  </StrokeButton>
</Playground>

## Toggleable

<Playground>
  <ToggleState>
    {({ on, toggle }) => (
      <StrokeButton showStroke={on} onClick={toggle}>
        Toggle me!
      </StrokeButton>
    )}
  </ToggleState>
</Playground>
