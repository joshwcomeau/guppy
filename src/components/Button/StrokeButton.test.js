/* eslint-disable flowtype/require-valid-file-annotation */
import React from 'react';
import { mount } from 'enzyme';
import DetectActive from '../DetectActive';
import StrokeButton, { Foreground, Background } from './StrokeButton';

describe('StrokeButton component', () => {
  let wrapper;
  let renderProp;
  beforeEach(() => {
    wrapper = mount(<StrokeButton />);
    renderProp = wrapper.find(DetectActive).renderProp('children');
  });

  it('should render foreground (inactive)', () => {
    expect(renderProp(false).find(Foreground)).toMatchSnapshot();
  });
  it('should render background (inactive)', () => {
    expect(renderProp(false).find(Background)).toMatchSnapshot();
  });

  it('should render foreground (active)', () => {
    expect(renderProp(true).find(Foreground)).toMatchSnapshot();
  });
  it('should render background (active)', () => {
    expect(renderProp(true).find(Background)).toMatchSnapshot();
  });
});
