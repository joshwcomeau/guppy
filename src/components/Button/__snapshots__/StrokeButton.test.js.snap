// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StrokeButton component should render background (active) 1`] = `
.c0 {
  position: absolute;
  z-index: 0;
  top: -2px;
  left: -2px;
  right: -2px;
  bottom: -2px;
  border-radius: 100px;
  background: linear-gradient(25deg,#651fff,#D500F9);
  opacity: 1;
  -webkit-transition: opacity 350ms;
  transition: opacity 350ms;
}

<styled.div
  colors={
    Array [
      "#651fff",
      "#D500F9",
    ]
  }
  isActive={true}
  isVisible={true}
>
  <div
    className="c0"
  />
</styled.div>
`;

exports[`StrokeButton component should render background (inactive) 1`] = `
.c0 {
  position: absolute;
  z-index: 0;
  top: 0px;
  left: 0px;
  right: 0px;
  bottom: 0px;
  border-radius: 100px;
  background: linear-gradient(25deg,#651fff,#D500F9);
  opacity: 1;
  -webkit-transition: opacity 350ms;
  transition: opacity 350ms;
}

<styled.div
  colors={
    Array [
      "#651fff",
      "#D500F9",
    ]
  }
  isActive={false}
  isVisible={true}
>
  <div
    className="c0"
  />
</styled.div>
`;

exports[`StrokeButton component should render foreground (active) 1`] = `
.c2 {
  position: relative;
  border: 0;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  background: #FFF;
  color: #111111;
  cursor: pointer;
  outline: none;
  white-space: nowrap;
}

.c2:disabled {
  -webkit-filter: grayscale(100%);
  filter: grayscale(100%);
  opacity: 0.75;
  cursor: initial;
}

.c2:not(:disabled):active {
  -webkit-transform-origin: center center;
  -ms-transform-origin: center center;
  transform-origin: center center;
}

.c1 {
  padding: 0px 20px;
  height: 38px;
  border-radius: 19px;
  font-size: 16px;
}

.c0 {
  display: block;
  position: relative;
  z-index: 1;
}

<styled.span>
  <span
    className="c0"
  >
    <ButtonBase
      background="#FFF"
      size="medium"
      textColor="#111111"
    >
      <Styled(styled.button)
        background="#FFF"
        textColor="#111111"
      >
        <styled.button
          background="#FFF"
          className="c1"
          textColor="#111111"
        >
          <button
            className="c1 c2"
          >
            <span
              style={
                Object {
                  "display": "block",
                }
              }
            />
          </button>
        </styled.button>
      </Styled(styled.button)>
    </ButtonBase>
  </span>
</styled.span>
`;

exports[`StrokeButton component should render foreground (inactive) 1`] = `
.c2 {
  position: relative;
  border: 0;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  background: #FFF;
  color: #111111;
  cursor: pointer;
  outline: none;
  white-space: nowrap;
}

.c2:disabled {
  -webkit-filter: grayscale(100%);
  filter: grayscale(100%);
  opacity: 0.75;
  cursor: initial;
}

.c2:not(:disabled):active {
  -webkit-transform-origin: center center;
  -ms-transform-origin: center center;
  transform-origin: center center;
}

.c1 {
  padding: 0px 20px;
  height: 38px;
  border-radius: 19px;
  font-size: 16px;
}

.c0 {
  display: block;
  position: relative;
  z-index: 1;
}

<styled.span>
  <span
    className="c0"
  >
    <ButtonBase
      background="#FFF"
      size="medium"
      textColor="#111111"
    >
      <Styled(styled.button)
        background="#FFF"
        textColor="#111111"
      >
        <styled.button
          background="#FFF"
          className="c1"
          textColor="#111111"
        >
          <button
            className="c1 c2"
          >
            <span
              style={
                Object {
                  "display": "block",
                }
              }
            />
          </button>
        </styled.button>
      </Styled(styled.button)>
    </ButtonBase>
  </span>
</styled.span>
`;
